crypto/fipsmodule/ec/ec_nistp.c

```diff
@@ -110,3 +122,135 @@ void ec_nistp_point_double(const ec_nistp_felem_meth *ctx,
+                        const ec_nistp_felem_limb *y1,
+                        const ec_nistp_felem_limb *z1,
+                        const int mixed,
+                        const ec_nistp_felem_limb *x2,
+                        const ec_nistp_felem_limb *y2,
+                        const ec_nistp_felem_limb *z2) {
+  ec_nistp_felem x_out, y_out, z_out;
+
+  ec_nistp_felem_limb z1nz = ctx->nz(z1);
+  ec_nistp_felem_limb z2nz = ctx->nz(z2);
+
+  // z1z1 = z1**2
+  ec_nistp_felem z1z1;
+  ctx->sqr(z1z1, z1);
+
+  ec_nistp_felem u1, s1, two_z1z2;
+  if (!mixed) {
+    // z2z2 = z2**2
+    ec_nistp_felem z2z2;
+    ctx->sqr(z2z2, z2);
+
+    // u1 = x1*z2z2
+    ctx->mul(u1, x1, z2z2);
+
+    // two_z1z2 = (z1 + z2)**2 - (z1z1 + z2z2) = 2z1z2
+    ctx->add(two_z1z2, z1, z2);
+    ctx->sqr(two_z1z2, two_z1z2);
+    ctx->sub(two_z1z2, two_z1z2, z1z1);
+    ctx->sub(two_z1z2, two_z1z2, z2z2);
+
+    // s1 = y1 * z2**3
+    ctx->mul(s1, z2, z2z2);
+    ctx->mul(s1, s1, y1);
+  } else {
+    // We'll assume z2 = 1 (special case z2 = 0 is handled later).
+
+    // u1 = x1*z2z2
+    OPENSSL_memcpy(u1, x1, ctx->felem_num_limbs * sizeof(ec_nistp_felem_limb));
+    // two_z1z2 = 2z1z2
+    ctx->add(two_z1z2, z1, z1);
+    // s1 = y1 * z2**3
+    OPENSSL_memcpy(s1, y1, ctx->felem_num_limbs * sizeof(ec_nistp_felem_limb));
+  }
+
+  // u2 = x2*z1z1
+  ec_nistp_felem u2;
+  ctx->mul(u2, x2, z1z1);
+
+  // h = u2 - u1
+  ec_nistp_felem h;
+  ctx->sub(h, u2, u1);
+
+  ec_nistp_felem_limb xneq = ctx->nz(h);
+
+  // z_out = two_z1z2 * h
+  ctx->mul(z_out, h, two_z1z2);
+
+  // z1z1z1 = z1 * z1z1
+  ec_nistp_felem z1z1z1;
+  ctx->mul(z1z1z1, z1, z1z1);
+
+  // s2 = y2 * z1**3
+  ec_nistp_felem s2;
+  ctx->mul(s2, y2, z1z1z1);
+
+  // r = (s2 - s1)*2
+  ec_nistp_felem r;
+  ctx->sub(r, s2, s1);
+  ctx->add(r, r, r);
+
+  ec_nistp_felem_limb yneq = ctx->nz(r);
+
+  // This case will never occur in the constant-time |ec_GFp_mont_mul|.
+  ec_nistp_felem_limb is_nontrivial_double =
+                                     constant_time_is_zero_w(xneq | yneq) &
+                                    ~constant_time_is_zero_w(z1nz) &
+                                    ~constant_time_is_zero_w(z2nz);
+  if (constant_time_declassify_w(is_nontrivial_double)) {
+    ec_nistp_point_double(ctx, x3, y3, z3, x1, y1, z1);
+    return;
+  }
+
+  // I = (2h)**2
+  ec_nistp_felem i;
+  ctx->add(i, h, h);
+  ctx->sqr(i, i);
+
+  // J = h * I
+  ec_nistp_felem j;
+  ctx->mul(j, h, i);
+
+  // V = U1 * I
+  ec_nistp_felem v;
+  ctx->mul(v, u1, i);
+
+  // x_out = r**2 - J - 2V
+  ctx->sqr(x_out, r);
+  ctx->sub(x_out, x_out, j);
+  ctx->sub(x_out, x_out, v);
+  ctx->sub(x_out, x_out, v);
+
+  // y_out = r(V-x_out) - 2 * s1 * J
+  ctx->sub(y_out, v, x_out);
+  ctx->mul(y_out, y_out, r);
+  ec_nistp_felem s1j;
+  ctx->mul(s1j, s1, j);
+  ctx->sub(y_out, y_out, s1j);
+  ctx->sub(y_out, y_out, s1j);
+
+  cmovznz(x_out, ctx->felem_num_limbs, z1nz, x2, x_out);
```

Comment:
It may be slightly more efficient to keep the original order as, for example, x_out would remain in registers/cache before sending it or x1 to x_3.

Reply:
I'm not sure it can make any difference in performance because: 1) the data is so small that everything is in cache anyway; 2) `cmovznz` is a function call so if not inlined it can't reuse values in registers. I'd like to keep this order for clarity if you don't mind?